### DISCUSSION TITLE
Example of synchronous http3 client without the need for manual IO?

### DISCUSSION THREAD

**Discussion by Szwendacz99** [aiortc/aioquic#500] [Q&A] [answered]
Hello,
I tried to write very basic client that would not use asyncio, but failed, due to the library and examples dependency of asynchronous approach (and maybe due to my lack of async programming experience). My case does not require performance, but rather easy tracking of the "connection", to for example easily tell which steps are taking longer than expected, or where exactly is the failure if there is any.

Is such synchronous client possible without implementing all the IO handling, `asyncio.DatagramTransport` etc.. ?

**Comment by rthalley** [comment, marked as answer]
It is possible, but you definitely have to implement the I/O.   It may not be the best example, but you could look at [dnspython's sync QUIC code](https://github.com/rthalley/dnspython/blob/main/dns/quic/_sync.py).

**Reply by Szwendacz99** [reply]
In the meantime I have modified the http3_client.py example so that I can measure required request steps times (and some more unrelated modifications) and that will be enough for me for now, but the example you pointed out will be very handy If I need to dig deeper in the protocols. 
Thanks.
